### PR TITLE
Feature/object oriented

### PR DESCRIPTION
**Pull request by zenflow**
I remembered a pattern I learned from @substack, that lets you to create prototypical classes that can be instantiated without the `new` keyword.. 

https://github.com/sweetalert2/sweetalert2/blob/a26d125ef75f137b288543296d304adbaa1d1a8f/src/sweetalert2.js#L81-L84

Using this we can enable a standard OO interface immediately, with almost full (see below) backwards compatibility.

So basically, you can still do `await swal('foo')`, but you can also do `await new swal('foo')`, and more importantly, `class extends swal { _main(params) { ... }}`. The swal instance has 1 public property: `params` and it is `Object.freeze`d. It's public methods are all the methods in `src/instanceMethods.js` (which are also still available as actually-static methods, until the next major release), plus `then()`, `catch()` and `finally()` to make it a thenable.

---

## Breaking change

The current changes as is will break any code depending on `swal('foo') instanceof Promise` being `true`, since `swal` will no longer return a `Promise` instance, but `Swal` instance which is thenable. It will be treated the same as a `Promise` in most cases, but there is the potential for this to break some expectations somewhere. I think it's probably negligible though, especially when considering the gain.

Another approach would be to **make** `swal('foo') instanceof Promise` be true, by just setting up the prototype chain, but that would only make it **look** like an instance of `Promise`, and could lead to more confusing bugs than the alternative of being just a simple thenable.

---

## Benefits

Just to make explicit:

Extensions can be defined concisely (i.e. currently there's a lot of boilerplate, especially when there's accompanying tests), and can be typed properly (see https://github.com/sweetalert2/sweetalert2-react-content/pull/15#issuecomment-372050612). 

...not to mention now we can think about what public instance methods/properties would make sense and be convenient. In the next major we can deprecate all the static methods that should be instance methods (and make them just instance methods). This change will also help in having a *better organized* API. Related: https://github.com/sweetalert2/sweetalert2.github.io/issues/15#issuecomment-374726441

---

### CODE REVIEW

**Comment by zenflow** [comment]
@limonte please review and let me know what you think, and then I have one more matter to discuss in regards to the `mixin` method.

**Comment by zenflow** [comment]
Check out https://github.com/sweetalert2/sweetalert2/pull/1036/commits/c4e7b7098cdfd154e3f97cd5745fec947af12137 and let me know if you want any more examples

**Comment by zenflow** [comment]
Here's another example of `extend`ing `Swal`... the `withCopyrightInFooter` example from #968 adapted:

```js
import Swal from 'sweetalert2'
export class CopyrightSwal extends Swal {
  _main (params) {
    return super._main({
      ...params,
      footer: (params.footer ? params.footer + '<br/>' : '') + 'Copyright Rick Sanchez'
    })
  }
}
```

or as a higher-order component, for better modularity and composibility:

```js
// example usage: `const CopyrightSwal = withCopyright(Swal)`
export function withCopyright (Swal) {
  return class extends Swal {
    _main (params) {
      return super._main({
        ...params,
        footer: (params.footer ? params.footer + '<br/>' : '') + 'Copyright Rick Sanchez'
      })
    }
  }
}
```

Now lets try extending `argsToParams`...

```js
class CopyrightSwal extends Swal {
  static argsToParams (args) {
    if (typeof args[3] === 'string') {
      // if the 4th arg is a string, make it the `footer` param...
      return {
        ...Swal.argsToParams(args.slice(0,3)),
        footer: args[3]
    } else {
      // otherwise just proxy to parent method
      return Swal.argsToParams(args)
    }
  }
  _main (params) {
    return super._main({
      ...params,
      footer: (params.footer ? params.footer + '<br/>' : '') + 'Copyright Rick Sanchez'
    })
  }
}
```

*Did you know*, `Swal.prototype._main` is an async (promise-returning) function?

```js
class MySwal extends Swal {
  async _main (params) {
    // do something with `params`
    let result = await super._main(params)
    // do something with `result`
    return result
  }
}
```

File: src/instanceMethods/_main.js [hunk 1]
```diff
@@ -0,0 +1,548 @@
+import {showWarningsForParams} from '../utils/params'
+import * as dom from '../utils/dom/index'
+import { swalClasses } from '../utils/classes'
+import { formatInputOptions, error, callIfFunction } from '../utils/utils'
+import setParameters from '../utils/setParameters'
+import globalState from '../globalState'
+import { openPopup } from '../utils/openPopup'
+
+export function _main (userParams) {
+  const ctor = this.constructor // dynamic reference to "the" (not necessarily our) constructor function
```
Comment (limonte) [inline src/instanceMethods/_main.js]: `ctor` is quite a unique variable name, `constructor` would be better IMO

Comment (zenflow) [reply src/instanceMethods/_main.js]: I defined `ctor` just to make the reference shorter, it's not really needed. If you want the ref to be `constructor` then we might as well just make it `this.constructor` which is probably better than either. Sound ok?

Comment (limonte) [reply src/instanceMethods/_main.js]: Yup :+1:

**Comment by zenflow** [comment]
@limonte ok the part about `mixin`.. I updated it to be OO using prototypical inheritance also, so that it is somewhat aligned with the new kind of extension, but pre-es6 oo patterns aren't 100% compatible with ES classes..

Currently, the following interoperation of `Swal.mixin` and `extends Swal`will not work...

```js
class ExtendedSwal extends Swal {}
const MixinSwal = ExtendedSwal.mixin({})
MixinSwal('foo') // <-- throws "Uncaught TypeError: Class constructor ExtendedSwal cannot be invoked without 'new'"
new MixinSwal('foo') // <-- throws the same error. the es super-constructor can not be `.apply`ed to the instance because `apply` can't signal the `new` keyword.
```

I suggest fixing it by making a small breaking change. Luckily the `.mixin` method is very new and not even documented yet. Let's make it use proper ES classes. That means that the class it returns *must* be instantiated with the `new` keyword. But it will allow us to mix `.mixin`s, and `extend`s together (specifically it will allow us to use `.mixin` after `extend`ing, which is the time when I'd want the fast-and-loose `mixin` method, not before).

If you would prefer to keep the ability to invoke mixin swals without the `new` keyword, I think we can handle that *(edit: we can't)*. I can modularize the main constructor method a wee bit more (i.e. extract a `doPreFlightCheck` function) and then duplicate the new tiny constructor in `mixin.js`.. that fixes the problem by avoiding the need to apply the parent constructor... ... ... oh but shit.. child classes can extend the constructor adding/changing logic, so it's not safe to assume we know what the parent constructor does..

**Comment by zenflow** [comment]
@limonte You ok with making that little breaking change to the `mixin` method?

**Comment by zenflow** [comment]
I think we should change all the tests and documentation to say `Swal` instead of `swal` and `SweetAlert` instead of `sweetAlert`, to reflect the new OO paradigm. Just change the function name, not use the `new` keyword. It could get confusing when we are talking about a hypothetical var named `swal` (instance) vs `Swal` (class).

Can I go ahead with this?

File: src/instanceMethods/_main.js (hunk 1 above)
Comment (zenflow) [reply src/instanceMethods/_main.js]: Done in https://github.com/sweetalert2/sweetalert2/pull/1036/commits/1a4928986a27f73e5557684ca64f650bca7661e0

**Comment by zenflow** [comment]
In https://github.com/sweetalert2/sweetalert2/pull/1036/commits/79aaea3d1b76be793bf0b663733b32651c2f5715 I just added a static `Swal.fire(...args)` method that simply creates and returns a `new` instance of itself, for use with `extend`ed `MySwal` classes, since es classes require you to use the `new` keyword, but linters complain about using the `new` keyword for side-effects:

e.g. 
```js
import Swal from 'sweetalert2'

class MySwal extends Swal {/* ... */}

new MySwal('foo', 'bar') // <-- Standard lint error: "Do not use 'new' for side effects."

MySwal.fire('foo', 'bar') // <-- do this instead 
```

Maybe in the next major we can "not use 'new' for side effects" and make the constructor a pure function by deprecating simply calling `await Swal(...args)`, and adding a `swal.run()` instance method that does the side-effects. Then `Swal.fire(...args)` would behave the same, but would be a shorthand for calling `(new Swal(...args).run()`.

**Comment by zenflow** [comment]
> I suggest fixing it by making a small breaking change. Luckily the .mixin method is very new and not even documented yet. Let's make it use proper ES classes. 

Done in https://github.com/sweetalert2/sweetalert2/pull/1036/commits/78661710869470aca5e9feae13e159803c1e6116

You must use `MySwal.fire(...args)` for the swals you create with `.mixin`.

**Comment by zenflow** [comment]
> I think we should change all the tests and documentation to say Swal instead of swal and SweetAlert instead of sweetAlert

Done in https://github.com/sweetalert2/sweetalert2/pull/1036/commits/6c2ee3730a38b221c923a86db81654e359f8a3a9 and https://github.com/sweetalert2/sweetalert2/pull/1036/commits/7fbce93061145e52b4344e2753cf3851e0c45b18

**Comment by zenflow** [comment]
Ready to merge IMO..

I'll follow up with a couple things
- separate logic for distinct features into HOCs, ready to be extracted into separate packages...
  - toasts
  - form-builder
  - global defaults, i.e. setDefaults/resetDefaults, which is deprecated but will be nice to have for legacy apps
  - "queue" which is kinda unnecessary too now that control-flow is so simple to do yourself with async/await. The nice thing about modularization like this is that we don't have to agree about the validity of features.. It can exist, and people can maintain it and use it, separately, without affecting the precious core.
  - maybe even the "input" feature?!?!

**Comment by zenflow** [comment]
I can't wait to integrate this change into sweetalert2-react-content

**Comment by limonte** [comment]
@zenflow I can't understand what is the reason of gzipped [dist size grown 0.5KB](https://bundlesize-store.now.sh/build?info=%7B%22files%22%3A%5B%7B%22maxSize%22%3A15360%2C%22path%22%3A%22dist%2Fsweetalert2.all.min.js%22%2C%22size%22%3A13813%2C%22compression%22%3A%22gzip%22%2C%22master%22%3A13252%7D%5D%2C%22repo%22%3A%22sweetalert2%2Fsweetalert2%22%2C%22branch%22%3A%22feature%2Fobject-oriented%22%2C%22commit_message%22%3A%22Merge%20aa72e5dceaaecbc3ff30d29b2f79451beab1fda6%20into%20e10ab3e496410219f17b8e470d073b046168c944%22%2C%22sha%22%3A%22aa72e5dceaaecbc3ff30d29b2f79451beab1fda6%22%7D)? In this PR there's just reorganization of the code, which shouldn't lead to the bigger dist.

**Comment by zenflow** [comment]
I tried e9692d6c0d026afa00961400edea4b97495ae964 but that barely made a difference

Also fixed a little problem here #1041, but I don't think that will make much of a difference either..

**Comment by zenflow** [comment]
@limonte ahh maybe this is it.. this is the compiled `mixin` function now (jsdoc comment excluded:

```js
var _extends$2 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function mixin(mixinParams) {
  var Swal = this;
  return function (_Swal) {
    _inherits(MixinSwal, _Swal);

    function MixinSwal() {
      _classCallCheck(this, MixinSwal);

      return _possibleConstructorReturn(this, (MixinSwal.__proto__ || Object.getPrototypeOf(MixinSwal)).apply(this, arguments));
    }

    _createClass(MixinSwal, [{
      key: "_main",
      value: function _main(params) {
        return _get(MixinSwal.prototype.__proto__ || Object.getPrototypeOf(MixinSwal.prototype), "_main", this).call(this, _extends$2({}, mixinParams, params));
      }
    }]);

    return MixinSwal;
  }(Swal);
}
```

**Comment by limonte** [comment]
This change is so good, I will merge it and make a new release right now!

**Pull request merged**